newIDE/app/src/UI/SortableVirtualizedItemList/index.js

```diff
@@ -95,6 +97,7 @@ class ItemsList extends React.Component<ItemsListProps, *> {
               selected={item === selectedItem}
               onItemSelected={this.props.onItemSelected}
               buildMenuTemplate={this.props.buildMenuTemplate}
+              hasErrors={errorItemsList.includes(item.getName())}
```

Comment:
Finally, here to `hasErrors={!!erroredItems[item.getName()]}`

Why? Because "includes" will do a search in the table of erroredItems.
Let's say we have 1000 resources, 50 items displayed on the screen, and 200 resources that are errored.

Searching in an array need to traverse the whole array, in the worst case. If we call N the length of the array, we say that the complexity is "N".
Here, our array of errors is 200 items long. So we're going to search in this array for all the 50 items display on screen. We're going to do 50*200 = 10 000 operations (in the worst case).

If instead we have an object, we can improve this. Why?
Because objects are stored internally as a "hashmap". Hashmaps are super great, because given a "key", they given you the value in constant time. (we say the complexity is 1)

Let's say we have 1000 resources, 50 items displayed on the screen, and 200 resources that are errored.
Now, our errors are stored in an object.
For all the 50 items, we're going to get the value for the key being the name of the resource. We're going to do 50*1 = 50 operations. Way faster! And if you have a lot of errors, it does not even matter, the number of operations stay constant, which is great.

Creating the hash map at the component mounting need to traverse the array of resources, which is done only once and very quick. The cost of creating this object ("hashmap") of errored items is so negligible and a great deal because it optimize a lot every other rendering :)

Reply:
If you're interested, it's a bit technical but search for "complexity of data structure" : http://blog.benoitvallon.com/data-structures-in-javascript/data-structures-in-javascript/
This looks like a gentle introduction to this field: https://medium.freecodecamp.org/time-is-complex-but-priceless-f0abd015063c

These are things that are a bit mathematical and more in the computer science field. You might not be interested, but still worth knowing a few fact.
Basically: searching in a big array = slow and searching/setting a value in an object = quick. The advantage of an array is that it's ordered... but here we don't care about the ordering of the errors ;) We just want to get them quick!

Reply:
@4ian this is gold! Thank you for taking the time to explain 😄 I am making the amendments now

Reply:
fantastic articles too. Thank you again for sharing this. I committed the changes now